File: tools/plan_vergleichen.py
"""
Druckt Personenplan und Bewegungsbild, wie der Server sie rechnet, zum Vergleich mit dem Browser.

Beides wird nirgends gepflegt, sondern aus den Ketten gerechnet — einmal im Browser für die
Ansicht, einmal hier für das Blatt in der Hand. Laufen die beiden auseinander, widerspricht der
Ausdruck dem Bildschirm, und genau das soll die Ableitung ja unmöglich machen.

    python tools/plan_vergleichen.py <arbeitsmappe.json>

`./toolchain.sh ablauf-vergleichen` ruft beide Seiten auf und vergleicht sie.
"""

import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent / "backend" / "src"))

from data.ablaufplan import plandaten  # noqa: E402
from entities.alarm import Arbeitsmappe  # noqa: E402

if len(sys.argv) < 2:
    sys.exit("Aufruf: python tools/plan_vergleichen.py <arbeitsmappe.json>")

mappe = Arbeitsmappe.model_validate(json.loads(Path(sys.argv[1]).read_text(encoding="utf-8")))

daten = plandaten(mappe)

for blatt in daten["personen"]:
    for zeile in blatt["zeilen"]:
        print(" | ".join([
            "PLAN", blatt["name"], zeile["datum"], zeile["von"], zeile["bis"], zeile["art"],
            zeile["vonOrt"], zeile["ort"], zeile["lage"],
            "faehrt" if zeile["faehrt"] else "-", zeile["fahrzeug"],
        ]))

for bild in daten["bewegung"]:
    print(" | ".join(["FENSTER", bild["datum"], str(bild["von"]), str(bild["bis"])]))
    for band in bild["baender"]:
        print(" | ".join(["BAND", bild["datum"], band["name"], str(band["reihen"])]))
    for balken in bild["balken"]:
        print(" | ".join([
            "BALKEN", bild["datum"], balken["schrittId"], balken["ortId"], str(balken["reihe"]),
            str(balken["von"]), str(balken["bis"]), balken["name"],
            ",".join(balken["besatzung"]), balken["lage"],
        ]))
    for linie in bild["linien"]:
        print(" | ".join([
            "LINIE", bild["datum"], linie["schrittId"], linie["vonOrtId"], str(linie["vonReihe"]),
            linie["nachOrtId"], str(linie["nachReihe"]), str(linie["von"]), str(linie["bis"]),
            linie["mittel"], linie["name"],
        ]))

File: tools/plan_vergleichen.mjs
/**
 * Druckt Personenplan und Bewegungsbild, wie der Browser sie rechnet, zum Vergleich mit dem
 * Server.
 *
 * Die Ableitung steht zweimal da — in TypeScript für die Ansicht, in Python für das PDF. Solange
 * beide dieselben Zeilen in derselben Reihenfolge liefern, kann der Ausdruck dem Bildschirm nicht
 * widersprechen; das ist der ganze Grund, aus dem beides überhaupt gerechnet und nicht gepflegt
 * wird.
 *
 *     node tools/plan_vergleichen.mjs <arbeitsmappe.json>
 */
import {readFileSync} from 'node:fs'
import {createRequire} from 'node:module'
import {dirname, resolve} from 'node:path'
import {fileURLToPath} from 'node:url'

const WURZEL = resolve(dirname(fileURLToPath(import.meta.url)), '..')
const holen = createRequire(resolve(WURZEL, 'frontend/package.json'))
const {build} = holen('esbuild')

const quelle = process.argv[2]
if (!quelle) {
    console.error('Aufruf: node tools/plan_vergleichen.mjs <arbeitsmappe.json>')
    process.exit(2)
}

async function laden(pfad) {
    const gebaut = await build({
        entryPoints: [resolve(WURZEL, pfad)],
        bundle: true, format: 'esm', write: false, platform: 'node',
    })
    return import('data:text/javascript;base64,' +
        Buffer.from(gebaut.outputFiles[0].text).toString('base64'))
}

const {personenplan} = await laden('frontend/src/scripts/ablauf.ts')
const {bewegungsbild, bewegungstage} = await laden('frontend/src/scripts/bewegungen.ts')

const mappe = JSON.parse(readFileSync(quelle, 'utf8'))
const daten = {planung: mappe.planung, fahrzeuge: mappe.kataloge?.fahrzeuge ?? []}

const namen = (liste, id, feld) => liste?.find(eintrag => eintrag.id === id)?.[feld] ?? ''
const ort = id => namen(mappe.planung.orte, id, 'name')
const lage = id => namen(mappe.planung.programmpunkte, id, 'name')
const fahrzeug = id => namen(mappe.kataloge?.fahrzeuge, id, 'funkrufname')

for (const person of mappe.planung.personen ?? []) {
    for (const eintrag of personenplan(daten, person.id)) {
        console.log([
            'PLAN', person.name,
            eintrag.schritt.von.slice(0, 10), eintrag.schritt.von.slice(11, 16),
            eintrag.schritt.bis.slice(11, 16), eintrag.schritt.art,
            ort(eintrag.vonOrtId), ort(eintrag.nachOrtId), lage(eintrag.schritt.programmpunktId),
            eintrag.faehrt ? 'faehrt' : '-', fahrzeug(eintrag.lauf.fahrzeugId),
        ].join(' | '))
    }
}

for (const tag of bewegungstage(daten)) {
    const bild = bewegungsbild(daten, tag)
    console.log(['FENSTER', bild.datum, bild.von, bild.bis].join(' | '))
    for (const band of bild.baender) {
        console.log(['BAND', bild.datum, band.name, band.reihen].join(' | '))
    }
    for (const balken of bild.balken) {
        console.log(['BALKEN', bild.datum, balken.schrittId, balken.ortId, balken.reihe,
                     balken.von, balken.bis, balken.name, balken.besatzung.join(','),
                     balken.lage].join(' | '))
    }
    for (const linie of bild.linien) {
        console.log(['LINIE', bild.datum, linie.schrittId, linie.vonOrtId, linie.vonReihe,
                     linie.nachOrtId, linie.nachReihe, linie.von, linie.bis, linie.mittel,
                     linie.name].join(' | '))
    }
}
